### PR TITLE
JAMES-3539 Implement PushSubscription/get

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala [hunk 1]
```diff
@@ -75,7 +75,7 @@ class PushSubscriptionProbeModule extends AbstractModule {
 }
 
 trait PushSubscriptionSetMethodContract {
-
+  private lazy val UTC_DATE_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssX")
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala:78]: it exists
`PushSubscriptionSetMethodContract.TIME_FORMATTER`

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala [hunk 2]
```diff
@@ -170,6 +170,306 @@ trait PushSubscriptionSetMethodContract {
...
+           |                        "types": ["Email"]
+           |                    }
+           |                ]
+           |            },
+           |            "c1"
+           |        ]
+           |    ]
+           |}""".stripMargin)
+  }
+
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala:472]: Should we need a test case filter properties should fail when the property is vertificationCode and "validated" is `false`?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JmapApi.scala [hunk 3]
```diff
@@ -88,7 +88,10 @@ class JMAPApi (methods: Set[Method], defaultCapabilities: Set[Capability]) {
         .fold(e => SFlux.just(InvocationWithContext(Invocation.error(ErrorCode.UnknownMethod, e.description, invocation.invocation.methodCallId), invocation.processingContext)),
           _ => SFlux.fromPublisher(method.process(capabilities, invocation, mailboxSession))))
       .getOrElse(SFlux.just(InvocationWithContext(Invocation.error(ErrorCode.UnknownMethod, invocation.invocation.methodCallId), invocation.processingContext)))
-      .onErrorResume(throwable => SMono.just(InvocationWithContext(Invocation.error(ErrorCode.ServerFail, throwable.getMessage, invocation.invocation.methodCallId), invocation.processingContext)))
+      .onErrorResume(throwable => {
+        throwable.printStackTrace()
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/JmapApi.scala:92]: `throwable.printStackTrace()` is it accepted?

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala (hunk 2 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala:470]: No need IMO.

verificationCode is always a valid property, but the underlying data is null if not yet validated.

**Review by vttranlina — APPROVED** [review]

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala [hunk 4]
```diff
@@ -196,6 +196,306 @@ trait PushSubscriptionSetMethodContract {
...
```
Comment (Arsnael) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala:498]: Following Tung's remark, should we have a test then showing we get the verification code property in the get when the sub has been validated?

Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala:498]: Oh yeah I wanted to add that test and likely forgot...

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala [hunk 5]
```diff
@@ -194,6 +199,357 @@ trait PushSubscriptionSetMethodContract {
            |}""".stripMargin)
   }
 
+  @Test
+  def getShouldReturnEmptyWhenNone(): Unit = {
```
[the remark this answers is not in the document]
Comment (Arsnael) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala:203]: Ah I missed that but it's a valid point actually... it's the setMethod contract here, not the getMethod one. Or we could change the contract name to `PushSubscriptionMethodContract`

[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/PushSubscriptionSetMethodContract.scala:203]: PushSubscriptionMethodContract :+1:

**Pull request merged**